### DISCUSSION TITLE
3-legged Oauth for my Twitter Python script

### DISCUSSION THREAD

**Discussion by KieranDL** [tweepy/tweepy#1758] [Questions] [answered]
I have the below basic python script and I would like to use the 3-legged Oauth that the Twitter API makes available to allow anyone to sign in with their Twitter account and use the script.

At the moment they can simply enter their Keys/Tokens to use it but I would like a standard Twitter user to be able to use it and not just a developer who knows these.

```
import tweepy

print('Please enter your Consumer Key')
consumer_key = input()
print('Please enter your Consumer Secret')
consumer_secret = input()
print('Please enter your Access Token')
key = input()
print('Please enter your Access Secret')
secret = input()

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(key, secret)
api = tweepy.API(auth, wait_on_rate_limit=True)

screen_name = auth

for follower in api.followers(screen_name): 
    print(follower.screen_name) 
```
Do I have to use my script in a web app to make this possible or is there something I can add to the script for this authorisation?

I would like it to work similarly to a third party app that you would grant access to get/post from your twitter.``

**Comment by stratospher** [comment]
What did you enter as the callback URL in twitter app settings?

I got it to work by making a server with Nodejs using [this](https://gist.github.com/moonmilk/035917e668872013c1bd) and setting the callback url in the twitter developer app to  http://localhost:3456.

**Comment by Harmon758** [comment, marked as answer]
I've rewritten the authentication documentation with https://github.com/tweepy/tweepy/commit/f9a722bae8cce368b9f8fd447c418e1034c32178, which should hopefully answer this question.

You should look into using [PIN-based OAuth](https://tweepy.readthedocs.io/en/v4.5.0/authentication.html#pin-based-oauth). Otherwise, you'll have to use a callback URL.

**Reply by SpangleLabs** [reply]
That PIN-based OAuth documentation seems to be wrong, I get:
```
AttributeError: 'OAuth1UserHandler' object has no attribute 'fetch_token'
```

I guess it should be Oauth2UserHandler? But that's constructed with different arguments and such

**Reply by SpangleLabs** [reply]
Ah, swapping it to OAuthHandler, [like this gist comment recommends](https://gist.github.com/moonmilk/035917e668872013c1bd?permalink_comment_id=3796711#gistcomment-3796711), and then tweaking some bits, works:

```
auth = tweepy.OAuthHandler(
    consumer_key, consumer_secret,
    callback="oob"
)
print(auth.get_authorization_url())
verifier = input("Input PIN: ")
auth.get_access_token(verifier)
access_token = auth.access_token
access_token_secret = auth.access_token_secret
```

**Reply by Harmon758** [reply]
This was a typo that should be fixed with #1805.
You should use [`OAuth1UserHandler.get_access_token`](https://tweepy.readthedocs.io/en/v4.5.0/authentication.html#tweepy.OAuth1UserHandler.get_access_token) instead.

[`OAuthHandler`](https://tweepy.readthedocs.io/en/v4.5.0/authentication.html#tweepy.OAuthHandler) is now a deprecated alias for [`OAuth1UserHandler`](https://tweepy.readthedocs.io/en/v4.5.0/authentication.html#tweepy.OAuth1UserHandler), and `get_access_token` returns the access token and secret directly.
